core/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -204,36 +209,44 @@ public void createSnapshot(final SnapshotRequest request, final CreateSnapshotLi
             public ClusterState execute(ClusterState currentState) {
                 validate(request, currentState);
 
+                // check if the snapshot name already exists in the repository
+                if (snapshotNameExists(repositoryName, snapshotName)) {
```

Comment:
This does IO in the cluster state update thread 👎 Super dangerous.
We should probably put assertions in BlobStoreRepository that we are not the cluster state update thread to help prevent mistakes like this.